### ISSUE TITLE
Use render hooks instead of email and img shortcodes

### ISSUE THREAD

**Issue by timtorres** [h-enk/hyas#183]
## Summary

Hello! I'm in the middle of migrating a project over to a Hyas starter & workflow and was merging in the Hyas shortcodes which include `email.html` and `img/-simple.html`. Shortcodes only work in Markdown and they only transform Markdown, so having used render hooks recently to augment default Markdown rendering while keeping the syntax, I was wondering why these shortcodes should be included/used at all.

### Motivation

- Shortcode syntax isn't supported by rich text editors unless you add a custom CMS object or are editing raw md
- When writing shortcode syntax manually it's difficult to remember the parameters and whether they're named or positional
- Render hooks can utilize partials which can be reused in templates and is more DRY (shortcodes can use partials too fwiw)
- You can now create type/section specific hooks for different behaviors, say image treatment/transformation on a blog vs product page
- Do we need to have the `img-simple` shortcode? What benefit does that bring? It would also be nice to document what the shortcodes (and other templates & code ) are doing in the documentation

### Basic example

Here's an example of my `render-link.html` which utilizes my `email.html` partial:
```
{{/* If the link is an email, use the Email partial to utilize email obfuscation. */}}
{{/* Opens external links in a new tab with a noopener ref, and downloadable files like .PDF or .zip (only, for now) */}}

{{- if (strings.HasPrefix .Destination "mailto") -}}
    {{ with .Text}}
        {{ partial "email" (dict "emailAddress" $.Destination "emailTitle" .) }}
    {{ else }}
        {{ partial "email" (dict "emailAddress" $.Destination) }}
    {{ end -}}
{{- else -}}
<a href="{{ .Destination | safeURL }}"{{ with .Title}} title="{{ . }}"{{ end }}{{ if (or (strings.HasPrefix .Destination "http") (strings.HasSuffix .Destination "pdf") (strings.HasSuffix .Destination "zip")) }} target="_blank" rel="noopener"{{ end }}>{{ .Text | safeHTML }}</a>
{{- end -}}
```

WIP `email.html`:
```
{{/* Encodes emails with JS to protect from bot scrapers. If the text attribute is included (like "Contact Tim") then display that in the anchor, otherwise display the email itself */}}
{{/* {{ partial "email" (dict "emailAddress" "email@address" "emailTitle" "Email Text") }} */}}
{{/* {{ partial "email" (dict "emailAddress" "email@address") }} */}}

{{- $address := replace (.emailAddress | safeURL) "mailto:" "" -}}
<script type="text/javascript">document.write("<a href='mailto:{{ $address }}'>" + "{{ with .emailTitle }}{{ . | safeHTML }}{{else}}{{ $address }}{{end}}" + "</a>");</script><noscript>{{ replace $address "@" " at " }}</noscript>
```

**Comment by h-enk** [comment]
@timtorres, thanks for this. I think there's no valid reason for using shortcodes in stead of render hooks. Will be looking into this ... soonish.

Input: [bep/portable-hugo-links](https://github.com/bep/portable-hugo-links)

**Comment by timtorres** [comment]
Just noting this affects the [blog archetype](https://github.com/h-enk/hyas/blob/master/archetypes/blog.md) (removing shortcode reference)

**Comment by h-enk** [comment]
>`document.write("<a href='mailto:{{ $address }}'>" + "{{ with .emailTitle }}{{ . | safeHTML }}{{else}}{{ $address }}{{end}}" + "</a>");`
 
This way, an email address (`$address`) will still be scrapable when viewing the page source.

**Comment by h-enk** [comment]
Implemented the image render hook (really nice, thanks for triggering this), but dropped the link render hook — see also https://github.com/h-enk/hyas/pull/188

**Comment by timtorres** [comment]
Great, nice one 👍  

>This way, an email address ($address) will still be scrapable when viewing the page source.

Ha yeah, this was still WIP as I explored some methods. Nothing's perfect but I found a good (but old) [breakdown](https://superuser.com/a/235965) of some different methods and your original and `noscript` fallback worked well enough for me as a balance between simplicity and effectiveness.

Replying to some of your notes from #188 here just to keep in context:

> Dropped link render hook containing "target="_blank" rel="noopener" logic ([G200](https://www.w3.org/TR/WCAG20-TECHS/G200.html): Opening new windows and tabs from a link only when necessary)

This makes sense when linking within the same or related domains, noting the exceptions provided, but isn't the general practice to open links leading to a new website in a new tab? If my users click on an external link then I'm losing them by directing them away from my site. I've rarely seen it done and I would rarely recommend it.

>Dropped link render hook containing email obfuscation (looks like it's not really possible). Kept email shortcode (for now, as an option, but better to avoid using email links)

I don't understand this one - the link render hook would do the same thing as the shortcode, whether or not obfuscation is effective, without having to use the shortcode. So better and easier workflow for editors. And markdown supports email links whether or not you're using a render hook, also making the shortcode unnecessary: `[email@domain.com](mailto:email@domain.com)` or `<email@domain.com>`
